### PR TITLE
Seeding and chaining optimizations

### PR DESCRIPTION
**Pull request by hmusta**
- [x] Filter out low complexity seeds using sdust
- [x] Optimize chainer: reduce memory overhead
- [x] AVX2 instructions in chainer
- [x] Simplify labeled seeding
- [x] Unit tests

### CODE REVIEW

File: metagraph/src/graph/alignment/aligner_chainer.cpp [hunk 1]
```diff
@@ -76,252 +104,460 @@ call_seed_chains_both_strands(std::string_view forward,
...
+    v = _mm256_castps_si256(_mm256_cvtepi32_ps(v)); // convert an integer to float
+    v = _mm256_srli_epi32(v, 23); // shift down the exponent
+    v = _mm256_subs_epu16(_mm256_set1_epi32(158), v); // undo bias
+    v = _mm256_min_epi16(v, _mm256_set1_epi32(32)); // clamp at 32
+
+    // log2
+    v = _mm256_sub_epi32(_mm256_set1_epi32(31), v);
+
+    return v;
+}
```
Comment (hmusta) [inline metagraph/src/graph/alignment/aligner_chainer.cpp:334]: this is redundant, remove it

File: metagraph/src/graph/alignment/dbg_aligner.cpp [hunk 2]
```diff
@@ -441,19 +457,22 @@ ::extend_chain(std::string_view query,
     assert(!best.empty());
 
     if (best.get_end_clipping()) {
+        logger->trace("Extending back");
```
Comment (karasikov) [inline metagraph/src/graph/alignment/dbg_aligner.cpp:460]: Would this not generate too much stuff? Maybe it's better to print it as `DEBUG(...)`?

File: metagraph/src/graph/alignment/dbg_aligner.cpp [hunk 3]
```diff
@@ -551,17 +577,10 @@ ::align_both_directions(std::string_view forward,
...
                         auto cur_columns = aln.label_columns;
                         if (!aggregator.add_alignment(std::move(aln))) {
-                            if (cur_columns.size()) {
-                                any_added = true;
-                                finished_columns.insert(cur_columns.begin(), cur_columns.end());
-                            }
-                        } else {
-                            any_added = true;
+                            finished_columns.insert(cur_columns.begin(),
+                                                    cur_columns.end());
```
Comment (karasikov) [inline metagraph/src/graph/alignment/dbg_aligner.cpp:581]:
```suggestion
                            finished_columns.insert(cur_columns.begin(), cur_columns.end());
```

File: metagraph/src/graph/alignment/alignment.hpp [hunk 4]
```diff
@@ -95,6 +95,47 @@ class Seed {
...
+            size_t clipping = transformer(it).get_clipping();
+            for (++it; it != end && transformer(it).get_offset() == offset
+                                    && transformer(it).get_clipping() == clipping; ++it) {}
+            --it;
+        }
+
+        last_q_end = q_end;
+    }
+    return num_matching;
+}
```
Comment (karasikov) [inline metagraph/src/graph/alignment/alignment.hpp:137]: If you always need first, why not just run `utils::get_first()` for all types?

File: metagraph/src/graph/alignment/aligner_seeder_methods.cpp [hunk 5]
```diff
@@ -201,6 +220,11 @@ void SuffixSeeder<BaseSeeder>::generate_seeds() {
         size_t seed_length = 0;
         std::vector<node_index> alt_nodes;
 
+        if (!this->config_.no_seed_complexity_filter &&
```
Comment (karasikov) [inline metagraph/src/graph/alignment/aligner_seeder_methods.cpp:223]: Why `no_`? Maybe it's better to call it `seed_complexity_filter` (though, the flag in CLI can still be `--no-...`)?

File: metagraph/src/graph/alignment/aligner_config.hpp [hunk 6]
```diff
@@ -51,6 +51,7 @@ struct DBGAlignerConfig {
     bool global_xdrop = true;
     bool allow_left_trim = true;
     bool no_backtrack = false;
+    bool no_seed_complexity_filter = false;
```
Comment (karasikov) [inline metagraph/src/graph/alignment/aligner_config.hpp:54]:
```suggestion
    bool seed_complexity_filter = true;
```

File: metagraph/src/graph/alignment/aligner_chainer.cpp [hunk 7]
```diff
@@ -1,63 +1,91 @@
 #include "aligner_chainer.hpp"
 
+#include <unordered_set>
```
Comment (karasikov) [inline metagraph/src/graph/alignment/aligner_chainer.cpp:3]: Why not `hopscotch_set`?

**Review by karasikov — CHANGES_REQUESTED** [review]
Could you also write in docs some guidelines for the choice of parameters?

File: metagraph/src/graph/alignment/aligner_chainer.cpp (hunk 7 above)
Comment (hmusta) [reply metagraph/src/graph/alignment/aligner_chainer.cpp:3]: I need it for `multiset`

File: metagraph/src/graph/alignment/aligner_seeder_methods.cpp (hunk 5 above)
Comment (hmusta) [reply metagraph/src/graph/alignment/aligner_seeder_methods.cpp:223]: I did it this way to make the variable have the same name as the CLI flag

File: metagraph/src/graph/alignment/aligner_chainer.cpp (hunk 7 above)
Comment (karasikov) [reply metagraph/src/graph/alignment/aligner_chainer.cpp:3]: Ok, if it's not critical for the performance, it's fine.

**Comment by hmusta** [comment]
I set up the unit tests for chaining, now I just need to add a few more to test corner cases.

File: metagraph/src/cli/query.cpp [hunk 8]
```diff
@@ -1141,6 +1142,20 @@ int query_graph(Config *config) {
...
+    if (auto dbg_succ = std::dynamic_pointer_cast<DBGSuccinct>(graph)) {
+        if (config->align_sequences && dbg_succ->get_mode() == DeBruijnGraph::PRIMARY) {
+            auto node_rc = std::make_shared<NodeRC>(*dbg_succ);
+            if (node_rc->load(config->infbase)) {
+                logger->trace("Loaded the adj-rc index (adjacent to reverse-complement nodes)");
+                dbg_succ->add_extension(node_rc);
+            } else {
+                logger->warn("adj-rc index missing or failed to load. "
+                             "Alignment speed will be significantly slower. "
+                             "Use metagraph transform to generate an adj-rc index.");
```
Comment (karasikov) [inline metagraph/src/cli/query.cpp:1154]:
```suggestion
                             "Run `metagraph transform --adj-rc ...` to generate the adj-rc index.");
```

File: metagraph/src/graph/alignment/aligner_seeder_methods.cpp [hunk 9]
```diff
@@ -1,5 +1,7 @@
 #include "aligner_seeder_methods.hpp"
 
+#include "sdust.h"
```
Comment (karasikov) [inline metagraph/src/graph/alignment/aligner_seeder_methods.cpp:3]:
```suggestion
#include <sdust.h>
```

**Review by karasikov — APPROVED** [review]
